graphqlmetrics/pkg/batch/processor.go

```diff
@@ -0,0 +1,180 @@
+	costFunc func(T) int,
+) *Processor[T] {
+	return &Processor[T]{
+		logger:       logger,
+		closeChan:    make(chan struct{}),
+		shutdownChan: make(chan struct{}),
+		config:       config,
+
+		queue:  make(chan T, config.MaxQueueSize),
+		buffer: make([]T, 0, defaultBufferSize),
+
+		processBatch: processFunc,
+		costFunc:     costFunc,
+	}
+}
+
+// Enqueue attempts to one or more elements to the queue. If the queue is full,
+// it will wait until the queue can accept more data or the context is cancelled.
+func (p *Processor[T]) Enqueue(ctx context.Context, element ...T) error {
+	for _, e := range element {
+		if !p.canAcceptBatch() {
+			return errors.New("processor is closed")
+		}
+
+		select {
+		case p.queue <- e:
+			p.logger.Debug("element added to the queue", zap.Any("element", e))
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	return nil
+}
+
+// Start starts the batch processor.
+func (p *Processor[T]) Start(ctx context.Context) {
+	p.logger.Debug("Starting processor")
+
+	ticker := time.NewTicker(p.config.Interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			p.logger.Debug("interval completed - processing batch")
+			p.process()
+		case element := <-p.queue:
+			p.logger.Debug("receiving element from queue",
+				zap.Any("element", element),
+				zap.Int("queue_size", len(p.queue)),
+			)
+
+			p.receiveElement(element)
+		case <-ctx.Done():
+			p.process()
+			p.Stop()
+		case <-p.closeChan:
+			close(p.shutdownChan)
+			return
+		}
+	}
+}
+
+// Stop instructs to processor to shut down.
+// It will ensure that all remaining items in the queue and buffer are processed.
+func (p *Processor[T]) Stop() {
+	p.logger.Info("Shutdown requested - stopping processor")
+	close(p.closeChan)
+	close(p.queue)
+
+	<-p.shutdownChan
+
+	p.logger.Info("processing remaining items in the queue", zap.Int("queue_size", len(p.queue)))
+	for batch := range p.queue {
+		p.receiveElement(batch)
+	}
+
+	p.logger.Info("processing remaining items in the buffer", zap.Int("buffer_size", len(p.buffer)))
+	if len(p.buffer) > 0 {
+		p.process()
+	}
+}
+
+func (p *Processor[T]) canAcceptBatch() bool {
+	select {
+	case <-p.closeChan:
+		return false
+	default:
+		return true
+	}
+}
+
+func (p *Processor[T]) process() {
+	p.logger.Debug("Processing batch", zap.Int("buffer_size", len(p.buffer)))
+	if err := p.processBatch(p.buffer); err != nil {
+		p.logger.Error("Failed to process batch", zap.Error(err))
+		return
+	}
+
+	p.buffer = p.buffer[:0]
+	p.currentBufferCost = 0
+
+	p.logger.Debug("Batch processed successfully")
+}
+
+// receiveElement receives a batch of buffer adds them to the buffer.
+// If the buffer is full, the batch processing will be invoked.
+func (p *Processor[T]) receiveElement(element T) {
+	p.buffer = append(p.buffer, element)
+
+	if p.costFunc != nil {
+		p.currentBufferCost += p.costFunc(element)
+	}
+
+	p.logger.Debug("current buffer cost", zap.Int("currentCost", p.currentBufferCost))
```

Comment:
Please use snake_case everywhere.

Reply:
the metrics_server wasn't using `snake_case` that's why I followed the existing naming. I will update everything

Reply:
Thank you